20240927-multiple-object-deletion.md

```diff
@@ -0,0 +1,158 @@
+tags: [ "gateway", "satellite", "satellite/metainfo", "uplink" ]
+---
+
+# Multiple Object Deletion
+
+## Essentials
+
+### Header
+
+Date: 2024-09-24
+
+Owner: Jeremy Wharton
+
+Accountable:
+- Edge team
+
+Consulted:
+- Product team
+
+Informed:
+- QA team
+
+### Context
+
+The satellite metainfo RPC service provides an API through which objects stored on the Storj network may be interacted with. While this service exposes an endpoint through which individual objects may be deleted, it lacks an endpoint for deleting multiple objects in a single request. This limitation impacts services that depend on the metainfo RPC service, such as our S3-compatible gateway, by requiring separate requests for each object in a bulk deletion operation.
+
+This approach has several issues:
+
+- Network overhead: The amount of time spent transmitting and processing network requests increases with the number of objects to be deleted. This not only delays bulk operations but also consumes unnecessary bandwidth.
+
+- Rate limiting: Each deletion request counts against the deletion rate limit of the project associated with the request. This reduces the amount of deletion operations that can be performed within a given time frame. Additionally, this may cause bulk deletions to fail partway if the rate limit is reached mid-operation.
+
+For further context, the gateway currently supports Amazon S3's [DeleteObjects](https://docs.aws.amazon.com/AmazonS3/latest/API/API_DeleteObjects.html) operation, which allows for the deletion of multiple objects with a single HTTP request. However, the lack of native support for bulk deletions in the metainfo RPC service makes this operation subject to the aforementioned issues, as the gateway must translate every S3 request to a series of RPC requests.
+
+### Goals
+
+The goal of the implementation described by this document is to allow multiple objects to be deleted in a single request to the satellite.
+
+### Approach / Design
+
+#### Satellite
+
+The satellite metainfo RPC service must be extended with an endpoint allowing for the deletion of multiple objects in a single request.
+
+Requests directed to this endpoint must contain the following information:
```

Comment:
S3 also supports something called _quiet mode_:

> The operation supports two modes for the response: verbose and quiet. By default, the operation uses verbose mode in which the response includes the result of deletion of each key in your request. In quiet mode the response includes only keys where the delete operation encountered an error. For a successful deletion in a quiet mode, the operation does not return any information about the delete in the response body.

Supporting this mode would allow us to save bandwidth on clients that use it.

Reply:
Good point; done.